src/rust/ide/lib/view/src/graph_editor/component/edge.rs

```diff
@@ -319,37 +779,274 @@ pub fn sort_hack_2(scene:&Scene) {
+                    vec![EdgeShape::id(&front.side_line),  EdgeShape::id(&back.side_line)  ],
+                    vec![EdgeShape::id(&front.corner),     EdgeShape::id(&back.corner)     ],
+                    vec![EdgeShape::id(&front.main_line),  EdgeShape::id(&back.main_line),
+                         EdgeShape::id(&front.arrow),      EdgeShape::id(&back.arrow)      ],
+                    vec![EdgeShape::id(&front.corner2),    EdgeShape::id(&back.corner2)    ],
+                    vec![EdgeShape::id(&front.side_line2), EdgeShape::id(&back.side_line2) ],
+                    vec![EdgeShape::id(&front.corner3),    EdgeShape::id(&back.corner3)    ],
+                    vec![EdgeShape::id(&front.port_line)                                   ],
+                ]
+            }
+        }
+    }
+
+    /// Return `Id`s that match the given index condition `cond`.
+    fn index_filtered_shapes<F:Fn(i32)-> bool>(&self, cond:F) -> Vec<display::object::Id> {
+        self.ordered_part_ids
+            .iter()
+            .enumerate()
+            .filter(|(index, _)| cond(*index as i32))
+            .flat_map(|(_index, ids)| ids.clone())
+            .collect()
+    }
+
+    /// Shapes that are output side of the split.
+    fn output_side_shapes(&self) -> Vec<display::object::Id> {
+        self.index_filtered_shapes(move |index| (index + 1) < self.split_index as i32)
+    }
+
+    /// Shapes that are input side of the split.
+    fn input_side_shapes(&self) -> Vec<display::object::Id> {
+        self.index_filtered_shapes(move |index| index > (self.split_index as i32 + 1))
+    }
+
+    /// Shapes that are at the split location and adjacent to it.
+    fn split_shapes(&self) -> Vec<display::object::Id> {
+        self.index_filtered_shapes(move |index| (index - self.split_index as i32).abs() <=1)
+    }
+}
+
+
+
 // ===========
 // === FRP ===
 // ===========
 
+/// FRP system that is used to collect and aggregate shape view events from the sub-shapes of an
+/// `Edge`. The Edge exposes the `mouse_down`/`mouse_over`/`mouse_out` streams, while the sub-shapes
+/// emit events via th internal `on_mouse_down`/`on_mouse_over`/`on_mouse_out` sources.
+#[derive(Clone,CloneRef,Debug)]
+#[allow(missing_docs)]
+pub struct ShapeViewEventsProxy {
+    pub mouse_down : frp::Stream,
+    pub mouse_over : frp::Stream,
+    pub mouse_out  : frp::Stream,
+
+    on_mouse_down : frp::Source<display::object::Id>,
+    on_mouse_over : frp::Source<display::object::Id>,
+    on_mouse_out  : frp::Source<display::object::Id>,
+}
+
+#[allow(missing_docs)]
+impl ShapeViewEventsProxy {
+    pub fn new(network:&frp::Network) -> Self {
+        frp::extend! { network
+            def on_mouse_over = source();
+            def on_mouse_out  = source();
+            def on_mouse_down = source();
+
+            mouse_down <- on_mouse_down.constant(());
+            mouse_over <- on_mouse_over.constant(());
+            mouse_out  <- on_mouse_out.constant(());
```

Comment:
the type of `on_mouse_down` is `()` so this constant doesn't make sense here. If you want to convert from `Source` to `Stream`, use `into` instead.

Reply:
`on_mouse_down` here is an `frp::Source<display::object::Id>`.